onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -2766,155 +2809,126 @@ Status TensorrtExecutionProvider::CreateNodeComputeInfoFromGraph(const GraphView
 
   if (!has_dynamic_shape) {
     std::string timing_cache_path = "";
-    bool engine_update = false;
     if (timing_cache_enable_) {
       timing_cache_path = GetTimingCachePath(global_cache_path_, compute_capability_);
     }
     {
-      // ifstream file check, engine serialization/deserialization and engine build are in critical section. It needs lock protection to prevent race condition when inferencing with multithreading.
+      // Engine serialization/deserialization and engine build are in critical section.
+      // It needs lock protection to prevent race condition when inferencing with multithreading.
       auto lock = GetApiLock();
 
-      // If explicit profile flag is on and engine cache enable flag is on,
-      // we need to compare explicit profiles and profiles used to build the engine in order to decide whether to rebuild the engine.
-      if (has_explicit_profile && engine_cache_enable_) {
-        engine_update = CompareProfiles(profile_cache_path, profile_min_shapes_, profile_max_shapes_, profile_opt_shapes_);
```

Comment:
The `CompareProfile()` should still be kept here for the case where the whole model is not TRT eligible (meaning there are partitions) and user specifies "explicit" shape range, right? So that TRT EP can still trigger engine rebuild when the shape ranges mismatch.

Reply:
> So that TRT EP can still trigger engine rebuild when the shape ranges mismatch.

I think with these new additions and mostly removals under this line, the engine rebuild would be always triggered in such a case, as per all the lines from 2829. So was your main point behind this comment to keep those two `if` flows:
`if (engine_cache_enable_ && !engine_decryption_enable_ && engine_file && !engine_update) {`
and:
`else if (engine_decryption_enable_ && engine_cache_enable_ && std::filesystem::exists(encrypted_engine_cache_path) && !engine_update) {`

so that engine deserialization would still be applicable in case of multiple partitions?

Reply:
I found it hard to review the code simply through GitHub since the logic of the entire `if (!has_dynamic_shape)` is modified quite a lot here 😂. You are right, the engine rebuild would be always triggered in such a case.

Then my question is what you mentioned as well.
In the case of multiple partitions, meaning `CreateNodeComputeInfoFromGraph` will always be called.
We still need to do engine deserialization if every subgraph has its corresponding engine cache.